mms_msg/sampling/pattern/meeting/state_based/sampler.py

```diff
@@ -0,0 +1,294 @@
+                When not set a uniform rng is used.
+            examples: (optional) List of all examples that are currently present in the meeting.
+            current_source: (optional) Source for which the overlap should be determined.
+
+        Returns: Sampled overlap
+        """
+        raise NotImplementedError
+
+
+class DistributionOverlapSampler(OverlapSampler):
+    """
+    Class which is used to sample overlap values for the generation of a meeting using a DistributionModel.
+    It is guaranteed that the given overlap values are valid and only a maximum number
+    of speakers is active simultaneously, when the sampler is called with ().
+    When the sampling fails a ValueError is returned.
+
+    Properties:
+        max_concurrent_spk: Maximum number of concurrent active speakers.
+        distribution: DistributionModel from which the overlap should be sampled.
+        hard_minimum_overlap: Hard minimum value for the overlap
+        hard_maximum_overlap: Hard maximum value for the overlap
+    """
+
+    def __init__(self, max_concurrent_spk: int, distribution: DistributionModel, hard_minimum_overlap: int = 0,
+                 hard_maximum_overlap: int = 1000000):
+        """
+        Args:
+            max_concurrent_spk: Maximum number of concurrent active speakers.
+            distribution: DistributionModel from which the overlap should be sampled.
+            hard_minimum_overlap: Hard minimum value for the overlap
+            hard_maximum_overlap: Hard maximum value for the overlap
+        """
+
+        self.max_concurrent_spk = max_concurrent_spk
+        self.distribution = distribution
+        self.hard_minimum_overlap = hard_minimum_overlap
+        self.hard_maximum_overlap = hard_maximum_overlap
+
+    def _sample_overlap(self, minimum_overlap: int, maximum_overlap: int,
+                        rng: np.random.random = np.random.default_rng(), examples: List[Dict] = None,
+                        current_source: Dict[str, Any] = None) -> int:
+        """
+        Internal function that samples overlap from the distribution with respect to the maximum
+        and minimum allowed overlap. examples and current_source are not used
+
+        Args:
+            maximum_overlap: Maximum for the overlap that is sampled
+            minimum_overlap: Minimum for the overlap that is sampled
+            rng: The numpy rng that should be used, the rng should generate a number in the interval [0,1).
+                When not set a uniform rng is used.
+            examples: (Not used in the implementation of this function)
+            current_source: (Not used in the implementation of this function)
+
+        Returns: Sampled overlap according to the distribution
+        """
+
+        return self.distribution.sample_value(rng, minimum_value=minimum_overlap, maximum_value=maximum_overlap)
+
+
+@dataclass
```

Comment:
Make sure that all dataclasses are frozen when possible (i.e., we want to guarantee that the parameters do not change during the sampling process)